### PR TITLE
Translation to zh_CN assigned to group 4

### PR DESCRIPTION
**Pull request by zhiiiyang**
The group 4 assignment is completed and ready for review.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
Thank you!! Pending review by @gaospecial

File: po/zh_CN.po [hunk 1]
```diff
@@ -495,228 +495,233 @@ msgstr ""
...
-msgstr ""
+msgstr "'new' 的长度是 %d。 它的长度必须和'which' (%d)的长度一致。"
 
 #: assign.c:1211
 #, c-format
 msgid ""
 "Item %d of 'which' is %d which is outside range of the length %d character "
 "vector"
 msgstr ""
+"'which' 的 %1%d 项是 %2%d 且超出了长度为%3%d字符的范围"
```
Comment (MichaelChirico) [inline po/zh_CN.po]: A bit awkward I know, but the correct way is `%1$d`, etc, I think that applies to a few of the others as well

Comment (zhiiiyang) [reply po/zh_CN.po]: I've just changed them all. Pushing them now. :)

File: po/zh_CN.po [hunk 2]
```diff
@@ -3056,11 +3061,11 @@ msgstr ""
 
 #: gsumm.c:47
 msgid "l is not an integer vector"
-msgstr "l �����������"
+msgstr "l 不是整数向量"
```
Comment (MichaelChirico) [inline po/zh_CN.po]: @gaospecial there's some issue (presumably encoding-related) leading to this (and everything following) to appear different even though it's not (this section was not part of grp 4 assignment).

feel free to ignore & focus on above.

Comment (zhiiiyang) [reply po/zh_CN.po]: Let me know if I need to do something from my end.

Comment (MichaelChirico) [reply po/zh_CN.po]: nothing for now. i'm looking at my local copy of the target `translation` branch and the issue seems the same. so it's not specific to your PR. maybe Chun-Hui's PR had the wrong encoding and went unnoticed...

**Comment by MichaelChirico** [comment]
Hey @gaospecial checking in again, please have a look here to review the translations when you get a chance 😃

File: po/R-zh_CN.po [hunk 3]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 "nomatch= must be either NA or NULL (or 0 for backwards compatibility which "
 "is the same as NULL)"
-msgstr ""
+msgstr "nomatch=必须NA(缺失值)或者是(NULL)空值(或者对于反向匹配来说是0，其实"
+"也是和NULL(空值)一样的"
 
 msgid "which= must be a logical vector length 1. Either FALSE, TRUE or NA."
-msgstr ""
+msgstr "which=必须是一个长度为一的逻辑向量，比如FALSE(假)，TRUE(真)或者是"
+"NA(缺失值)"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: which= 必须是一个长度为 1 的逻辑向量。其取值为 FALSE，TRUE 或者 NA。

File: po/R-zh_CN.po [hunk 4]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
 msgstr "'(字符类型)。 唯一有效的字符值是'nearest'。"
 
 msgid "rollends must be a logical vector"
-msgstr ""
+msgstr "rollends必须是一个逻辑向量"
 
 msgid "rollends must be length 1 or 2"
-msgstr ""
+msgstr "rollends必须是长度为一或者是二"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: rollends 的长度必须是 1 或者 2

File: po/R-zh_CN.po [hunk 5]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 msgid "which=="
-msgstr ""
+msgstr "which=="
 
 msgid ""
 "(meaning return row numbers) but j is also supplied. Either you need row "
 "numbers or the result of j, but only one type of result can be returned."
-msgstr ""
+msgstr "(表示行数会被返回) 但是j也被提供。你需要行数或者是j的结果，但是只有"
+"一种结果会被返回。"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: (表示行数会被返回) 但是 j 也被提供了。你可能需要行数或者是 j 的结果，但是只能返回一种结果。

File: po/R-zh_CN.po [hunk 6]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 
 msgid ""
 "which=NA with nomatch=0 would always return an empty vector. Please change "
 "or remove either which or nomatch."
-msgstr ""
+msgstr "如果同时使用which=NA(缺失值)和nomatch=0会导致空向量的产生，所以请"
+"改变或者是移除which或者是nomatch的赋值"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: 同时使用 which=NA 和 nomatch=0 会得到一个空向量。请改变或者是移除 which 或 nomatch 的取值

File: po/R-zh_CN.po [hunk 7]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 
 msgid "j must be provided when with=FALSE"
-msgstr ""
+msgstr "如果with=FALSE(假)，j必须要赋值"
 
 msgid ""
 "The symbol .. is invalid. The .. prefix must be followed by at least one "
 "character."
-msgstr ""
+msgstr "这个符号..是无效的。紧接着前缀..的后面必须至少要有一个字符"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: 符号 .. 是无效的。前缀 .. 之后必须要有至少一个字符

File: po/R-zh_CN.po [hunk 8]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
+"with=FALSE。请使用 .. 符号前缀并且移除 with=FALSE。"
 
 msgid ""
 "You have wrapped := with {} which is ok but then := must be the only thing "
 "inside {}. You have something else inside {} as well. Consider placing the "
 "{} on the RHS of := instead; e.g. DT[,someCol:={tmpVar1<-...;tmpVar2<-...;"
 "tmpVar1*tmpVar2}"
-msgstr ""
+msgstr "用{}套用:=暂时可行，但是{}中只能包含:=。你现在{}有其他的参数。请尝试将"
+"使用:=的RHS；比如，DT[,someCol:={tmpVar1<-...;tmpVar2<-...;tmpVar1*tmpVar2}"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: 用 {} 套用 := 可行，但是 {} 中只能有 := 。而你现在 {} 中还有其他的参数。请尝试将
{} 置于 := 的RHS之上；比如，DT[,someCol:={tmpVar1<-...; tmpVar2<-...; tmpVar1*tmpVar2}

File: po/R-zh_CN.po [hunk 9]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 msgid ""
 ":= with keyby is only possible when i is not supplied since you can't setkey "
 "on a subset of rows. Either change keyby to by or remove i"
-msgstr ""
+msgstr ":=和keyby的组合只有在i没有赋值下才合理存在。因为你不能在一个行的子集"
+"调用setkey。要么把keyby换成by或者是移除i"
 
 msgid "nomatch isn't relevant together with :=, ignoring nomatch"
-msgstr ""
+msgstr "nomatch并不和:=有任何的相关，请忽略nomatch"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: nomatch 并不和 := 有任何的相关，将忽略nomatch

File: po/R-zh_CN.po [hunk 10]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 
 msgid ""
 "not-join '!' prefix is present on i but nomatch is provided. Please remove "
 "nomatch."
-msgstr ""
+msgstr "不能连接在i的'!'前缀，但是有提供nomatch。请移除nomatch。 "
```
Comment (gaospecial) [inline po/R-zh_CN.po]: not-join "!" 前缀在 i 中存在，但是 nomatch 也被提供了。需要移除 nomatch。

File: po/R-zh_CN.po [hunk 11]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 
 msgid ""
 "i is invalid type (matrix). Perhaps in future a 2 column matrix could return "
 "a list of elements of DT (in the spirit of A[B] in FAQ 2.14). Please report "
 "to data.table issue tracker if you'd like this, or add your comments to FR "
 "#657."
-msgstr ""
+msgstr "i不是一个有效的类型(矩阵)。也许在以后一个包含两列的矩阵会返回"
+"包含一串元素的DT (请参考问答集2.14的A[B])。如果你有需求，请将此问题汇报给"
+"data.table 问题追踪器或者是在FR中留下你的想法"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: "data.table issues 或者是在FR中留下你的想法"

File: po/R-zh_CN.po [hunk 12]
```diff
@@ -277,122 +277,135 @@ msgid "' (type character). Only valid character value is 'nearest'."
...
 "When i is a data.table (or character vector), the columns to join by must be "
 "specified using 'on=' argument (see ?data.table), by keying x (i.e. sorted, "
 "and, marked as sorted, see ?setkey), or by sharing column names between x "
 "and i (i.e., a natural join). Keyed joins might have further speed benefits "
 "on very large data due to x being sorted in RAM."
-msgstr ""
+msgstr "但i是一个data.table(或者是字符向量)，列的连接必须要先申明'on='的参数"
+"(请参考 ?data.table)，或者是keying x(比如，已排序过，和标记已排序过，请参考"
+"?setkey)，或者是在x和i共用列的名字(比如，自然连接)。如果x有在内存被排序过，"
+"Keyed连接的速度会在非常大的数据上有较明显的提高。"
```
Comment (gaospecial) [inline po/R-zh_CN.po]: msgstr "当 i 是一个 data.table（或字符向量），必须使用 'on=' 参数指明参与连接的列（参见 ?data.table）。"
"可以是keying x(比如，已排序过，和标记为已排序过，参见 ?setkey)，或者是在x和i共用列的名字(比如，自然连接)。"
"如果x有在内存被排序过，那么Keyed连接的速度会在数据比较大的时候有明显的提高。"

File: po/zh_CN.po [hunk 13]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
 "data.table issue tracker."
 msgstr ""
+"内部错误：savetl_init的校验失败 (%d %d %p %p)，请将此问题汇报给data.table 问"
+"题追踪器。"
 
 #: assign.c:1160
 #, c-format
 msgid "Failed to allocate initial %d items in savetl_init"
-msgstr ""
+msgstr "不能给 savetl_init 最开始的 %d 个项进行分配 "
```
Comment (gaospecial) [inline po/zh_CN.po]: 不能为 savetl_init 最开始的 %d 个项分配空间

File: po/zh_CN.po [hunk 14]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
 
 #: assign.c:1169
 #, c-format
 msgid ""
 "Internal error: reached maximum %d items for savetl. Please report to data."
 "table issue tracker."
 msgstr ""
+"内部错误：已经达到了savetl能处理的最多 %d 数量的子项。请将此问题汇报给data."
```
Comment (gaospecial) [inline po/zh_CN.po]: 内部错误：已经达到了 savetl 能处理的子项上限 %d 。请将此问题汇报给data.

File: po/zh_CN.po [hunk 15]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
+"table 问题追踪器。"
 
 #: assign.c:1176
 #, c-format
 msgid "Failed to realloc saveds to %d items in savetl"
-msgstr ""
+msgstr "不能给savetl里的 %d 个项提供 saveds"
```
Comment (gaospecial) [inline po/zh_CN.po]: 不能给 savetl 里的 %d 个项重新分配 saveds

File: po/zh_CN.po [hunk 16]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
 
 #: assign.c:1182
 #, c-format
 msgid "Failed to realloc savedtl to %d items in savetl"
-msgstr ""
+msgstr "不能给savetl里的 %d 个项提供 savetl"
```
Comment (gaospecial) [inline po/zh_CN.po]: 不能给 savetl 里的 %d 个项重新分配 saveds

File: po/zh_CN.po [hunk 17]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
-msgstr ""
+msgstr "'new' 的长度是 %d。 它的长度必须和'which' (%d)的长度一致。"
 
 #: assign.c:1211
 #, c-format
 msgid ""
 "Item %d of 'which' is %d which is outside range of the length %d character "
 "vector"
-msgstr ""
+msgstr "'which' 的 %1$d 项是 %2$d 且超出了长度为%3$d字符的范围"
```
Comment (gaospecial) [inline po/zh_CN.po]: 'which' 的 %d 项是 %d，这超出了 %d 字符的长度范围

File: po/zh_CN.po [hunk 18]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
 #, c-format
 msgid "Internal error: dt passed to setcolorder has %d columns but %d names"
-msgstr ""
+msgstr "内部错误: setcolorder读取到的dt有 %d 列但是有 %d 个名字。"
 
 #: assign.c:1230
 msgid ""
 "Internal error: o passed to Csetcolorder contains an NA or out-of-bounds"
-msgstr ""
+msgstr "内部错误: Csetcolorder读取到的o有一个NA(缺失值)或者是出界"
```
Comment (gaospecial) [inline po/zh_CN.po]: 内部错误: Csetcolorder读取到的o有一个NA(缺失值)或者是下标出界

File: po/zh_CN.po [hunk 19]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
+msgstr "内部错误: Csetcolorder读取到的o含有一个重复值"
 
 #: between.c:12
 #, c-format
 msgid ""
 "Incompatible vector lengths: length(x)==%d length(lower)==%d length(upper)=="
 "%d. Each should be either length 1 or the length of the longest."
 msgstr ""
+"向量长度不一致：length(x)==%d length(lower)==%d length(upper)==%d。每一个向量"
+"要么是长度为一或者是最长的长度。"
```
Comment (gaospecial) [inline po/zh_CN.po]: 要么是长度为一或者是等于最长向量的长度。

File: po/zh_CN.po [hunk 20]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
 #: between.c:22
 msgid "check must be TRUE or FALSE"
-msgstr ""
+msgstr "check 必须是 TRUE (真) 或者是 FALSE (假)"
 
 #: between.c:64
 #, c-format
 msgid "Item %d of lower (%d) is greater than item %d of upper (%d)"
-msgstr ""
+msgstr "lower (%1$d) 的 %2$d 项大于 upper (%3$d)的 %4$d 项"
```
Comment (gaospecial) [inline po/zh_CN.po]: lower (%2$d) 的 %1$d 项大于 upper (%3$d)的 %4$d 项

File: po/zh_CN.po [hunk 21]
```diff
@@ -545,229 +545,237 @@ msgstr "请使用 as.character() 把 integer64 类型的数值赋值给字符列
...
 #, c-format
 msgid "between parallel processing of integer64 took %8.3fs\n"
-msgstr ""
+msgstr "64位整数的并行运行时间一共花费了 %8.3fs\n"
 
 #: between.c:115
 msgid ""
 "x is not integer64 but lower and/or upper is integer64. Please align classes."
 msgstr ""
+"x 不是一个64位整数但是至少在 lower 和 upper 其中有一个是64位整数，请保证所有"
```
Comment (gaospecial) [inline po/zh_CN.po]: x 不是一个64位整数但是 lower 和 upper 中至少有一个是64位整数，请保证所有

**Comment by MichaelChirico** [comment]
@gaospecial I suspect you clicked on a specific commit's diff in an email?

that was from a merge commit; all the files are still shown on the `Changes` tab.

it's not made very prominent/visually obvious, but there's usually a link about "this is a subset of changes, click here for the full changes"

File: po/R-zh_CN.po (hunk 11 above)
Comment (zhiiiyang) [reply po/R-zh_CN.po]: I used 问题追踪器 to be consistent with others.

File: po/zh_CN.po (hunk 17 above)
Comment (zhiiiyang) [reply po/zh_CN.po]: I keep the number before $d to be clear but I agree that they are the same.

**Comment by MichaelChirico** [comment]
Thanks so much @zhiiiyang and @gaospecial !!

**Pull request merged**